### PR TITLE
Fix a racing condition when subscribing to a completed ScopeProvider.

### PR DESCRIPTION
**Pull request by duanbo1983**
**Description**:

**Related issue(s)**:

### CODE REVIEW

**Comment by duanbo1983** [comment]
Let me switch an account.

File: autodispose/src/main/java/com/uber/autodispose/AutoDisposingObserverImpl.java [hunk 1]
```diff
@@ -55,6 +55,8 @@
 
           @Override public void onComplete() {
             callMainSubscribeIfNecessary(d);
```
Comment (duanbo1983) [inline autodispose/src/main/java/com/uber/autodispose/AutoDisposingObserverImpl.java]: When subscribing to a completed ScopeProvider, callMainSubscribeIfNecessary() might be called here before the main source is subscribed, which triggers the DoubleSubscriptoinsException.

File: autodispose/src/test/java/com/uber/autodispose/AutoDisposeObserverTest.java [hunk 2]
```diff
@@ -171,6 +177,24 @@
     assertThat(lifecycle.hasObservers()).isFalse();
   }
 
+  @Test public void autoDispose_withScopeProviderCompleted_shouldNotReportDoubleSubscriptions() {
+    RxJavaPlugins.setErrorHandler(new Consumer<Throwable>() {
+      @Override
+      public void accept(Throwable throwable) throws Exception {
+        assertThat(throwable instanceof ProtocolViolationException).isFalse();
```
Comment (duanbo1983) [inline autodispose/src/test/java/com/uber/autodispose/AutoDisposeObserverTest.java]: This test case fails in master as:
`Exception in thread "main" io.reactivex.exceptions.ProtocolViolationException: It is not allowed to subscribe with a(n) com.uber.autodispose.AutoDisposingObserverImpl multiple times. Please create a fresh instance of com.uber.autodispose.AutoDisposingObserverImpl and subscribe that to the target source instead.
	at com.uber.autodispose.AutoDisposeEndConsumerHelper.reportDoubleSubscription(AutoDisposeEndConsumerHelper.java:110)
	at com.uber.autodispose.AutoDisposeEndConsumerHelper.setOnce(AutoDisposeEndConsumerHelper.java:56)
	at com.uber.autodispose.AutoDisposingObserverImpl.onSubscribe(AutoDisposingObserverImpl.java:67)
	at io.reactivex.subjects.PublishSubject.subscribeActual(PublishSubject.java:85)
	at io.reactivex.Observable.subscribe(Observable.java:10842)
	at com.uber.autodispose.ObservableScoper$AutoDisposeObservable.subscribeActual(ObservableScoper.java:113)
	at io.reactivex.Observable.subscribe(Observable.java:10842)
	at com.uber.autodispose.ObservableScoper$1.subscribe(ObservableScoper.java:94)
	at com.uber.autodispose.AutoDisposeObserverTest.autoDispose_withScopeProviderCompleted_shouldNotReportDoubleSubscriptions(AutoDisposeObserverTest.java:192)`

File: autodispose/src/main/java/com/uber/autodispose/AutoDisposingObserverImpl.java [hunk 3]
```diff
@@ -46,15 +46,20 @@
           @Override public void onSuccess(Object o) {
             callMainSubscribeIfNecessary(d);
             AutoDisposingObserverImpl.this.dispose();
```
Comment (hzsweers) [inline autodispose/src/main/java/com/uber/autodispose/AutoDisposingObserverImpl.java]: This should be changed to `AutoDisposableHelper.dispose(mainDisposable)`

File: autodispose/src/test/java/com/uber/autodispose/AutoDisposeObserverTest.java (hunk 2 above)
Comment (hzsweers) [reply autodispose/src/test/java/com/uber/autodispose/AutoDisposeObserverTest.java]: I don't quite understand this test, as the `ProtocolViolationException` isn't obvious. Some sort of comment and asserting what you do expect should happen here

It also won't fail if this callback isn't hit, let's make it explicit somehow that we expect this callback to be hit

**Review by ZacSweers — CHANGES_REQUESTED** [review]
Good start. Can you also update the other observers? i.e. for `Single`, `Maybe`, `Completable`, and `Flowable`

**Comment by hzsweers** [comment]
Talked offline, closing in favor of #138

**Pull request closed without merging**